site/source/docs/porting/simd.rst

```diff
@@ -3,27 +3,76 @@
+An upcoming `Relaxed SIMD proposal <https://github.com/WebAssembly/relaxed-simd/tree/main/proposals/relaxed-simd>`_ will add more SIMD instructions to WebAssembly.
+
+================================
+GCC/Clang SIMD Vector Extensions
+================================
+
+At the source level, the GCC/Clang `SIMD Vector Extensions <https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html>`_ can be used and will be lowered to WebAssembly SIMD instructions where possible.
+
+This enables developers to create custom wide vector types via typedefs, and use arithmetic operators (+,-,*,/) on the vectorized types, as well as allow individual lane access via the vector[i] notation. However, the `GCC vector built-in functions <https://gcc.gnu.org/onlinedocs/gcc/x86-Built-in-Functions.html>`_ are not available. Instead, use the WebAssembly SIMD Intrinsics functions below.
+
+===========================
+WebAssembly SIMD Intrinsics
+===========================
+
+LLVM maintains a WebAssembly SIMD Intrinsics header file that is provided with Emscripten, and adds type definitions for the different supported vector types.
 
     .. code-block:: cpp
 
        #include <wasm_simd128.h>
+       #include <stdio.h>
+
+       int main() {
+       #ifdef __wasm_simd128__
+         __f32x4 v1 = wasm_f32x4_make(1.2f, 3.4f, 5.6f, 7.8f);
+         __f32x4 v2 = wasm_f32x4_make(2.1f, 4.3f, 6.5f, 8.7f);
+         __f32x4 v3 = v1 + v2;
+         // Prints "v3: [3.3, 7.7, 12.1, 16.5]"
+         printf("v3: [%.1f, %.1f, %.1f, %.1f]\n", v3[0], v3[1], v3[2], v3[3]);
+       #endif
+       }
 
-Separate documentation for the intrinsics header is a work in progress, but its usage is straightforward and its source can be found at `wasm_simd128.h <https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/wasm_simd128.h>`_. These intrinsics are under active development in parallel with the SIMD proposal and should not be considered any more stable than the proposal itself. Note that most engines will also require an extra flag to enable SIMD. For example, Node requires `--experimental-wasm-simd`.
+The Wasm SIMD header can be browsed online at `wasm_simd128.h <https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/wasm_simd128.h>`_.
 
-WebAssembly SIMD is not supported when using the Fastcomp backend.
+Pass flag ``-msimd128`` at compile time to enable targeting WebAssembly SIMD Intrinsics. C/C++ code can use the built-in preprocessor define ``#ifdef __wasm_simd128__`` to detect when building with WebAssembly SIMD enabled.
+
+Pass ``-mrelaxed-simd`` to target WebAssembly Relaxed SIMD Intrinsics. C/C++ code can use the built-in preprocessor define ``#ifdef __wasm_relaxed_simd__`` to detect when this target is active. At the time of writing, Relaxed SIMD is experimental.
```

Comment:
```suggestion
Pass ``-mrelaxed-simd`` to target WebAssembly Relaxed SIMD Intrinsics. C/C++ code can use the built-in preprocessor define ``#ifdef __wasm_relaxed_simd__`` to detect when this target is active.
```
Relaxed SIMD is at phase 4 (although there seems to be some confusion about this) and shipped in Chrome 114, so I wouldn't call it experimental anymore.

Reply:
gotcha, removed that statement.

Reply:
Btw, I was eyeing at relaxed SIMD to figure out if those could be used to implement better SSE/NEON headers. But it looks like that might not be possible, since the relaxed SIMD instructions do not provide SSE/NEON semantical guarantees?

Reply:
Has it btw been considered to add instructions that would always behave like x86 semantics (and ARM targets would need to emulate) or instructions that would always behave like ARM semantics (and x86 targets would need to emulate)?

Such instructions would have the benefit of
a) being fully defined (no "implementation defined" business),
b) helping porting existing SSE and NEON code to be more performant (a more direct access to hw instructions, avoids the double emulation problem),
c) helping avoid fingerprinting possibilities (implementations cannot just compute functions with corner cases to dig out ARM vs x86 architectures, but at minimum they'll have to time them)

Reply:
We've definitely talked about it, but there hasn't been a serious proposal in that direction yet. The lack of portable performance would be a sticking point in the community group, although you're right that there would be a lot of benefits, too.

Reply:
Thanks, I follow parts of that argument, although

> The lack of portable performance would be a sticking point

I have long tried to argue that is a fallacy. In my view, the situation with adding fixed semantic 1:1 SSE/NEON opcodes would be completely the opposite and strictly better for all CPU hardware, with more portable performance.

If something, it is the status today that the existing Wasm SIMD instructions have a lack of portable performance.

The reason I think so is: today the instructions already are not 100% guaranteed to be 1:1 mappings to native SSE&NEON intrinsics, but some amount of instruction combinations may be performed by VMs to implement the opcodes.

In past years, I've asked for the spec to provide performance guarantees about what this "some amount" is, but got shot down on such requests on multiple grounds (too hard to spec, different VMs do different things, JIT, AOT vs interpreting tiers, etc). The situation today is in a state of a lack of portable performance.

Today, if I have an application that \*needs\* to calculate an opcode with, say, exact SSE semantics that doesn't exist in Wasm SIMD, I'll have to emulate SSE->WasmSIMD (either by manually rewriting the algo, or using xmmintrin.h, doesn't matter, both have same result: I get more instructions), and then the VM will emulate back WasmSIMD->SSE or WasmSIMD->NEON.

Either way, there will be *\*two layers\** of added instructions and I can never achieve the native performance, neither on SSE or NEON CPUs.

(can switch SSE<->NEON in above argument symmetrically)

Instead, if WasmSIMD provided that direct opcode for precise SSE semantics computation, the result would be \**strictly\** better for \**both\** VMs running on top of SSE and NEON hardware. That is because on SSE the result would be a guaranteed 1:1 opcode mapping as SSE->WasmSIMD_SSE would be a direct match; but also for a VM running on NEON hardware, there would now be *\*only one\** layer of mapping: WasmSIMD_SSE->NEON, which is strictly better than the two layers from before.

I.e. everyone strictly wins. It would not be a situation of "favoring SSE and pessimizing NEON" performance.

My impression is that the reason why people never wanted to buy the above argument is that
a) they might not think holistically enough about the computational problems that are being solved with SIMD, or
b) they view that all algorithms would/should be rewritable to naturally use only the SSE ∩ Wasm subset of SIMD opcode instructions, and programmer has not tried hard enough if not,
c) they feel too principled to add multiple variants of same SIMD functions into the API, a SSE ∪ Wasm API would look ugly
d) they outright don't like SSE, and want to favor NEON as a more thought out API.

(or at least that's the kind of mix of conversations I've had about this in the past near-decade)

I think Relaxed SIMD took a bit of a unfortunate direction in design. In my view it would have been better for Relaxed SIMD to evolve the spec from SSE ∩ Wasm more over to SSE ∪ Wasm, but it instead got these implementation defined instructions, that are difficult to rely on. For example the laneselect instruction is puzzling: what the whole instruction does is implementation defined, and not just corner cases that one might not care about. (https://github.com/WebAssembly/relaxed-simd/issues/152)